batch/batch/cloud/gcp/driver/pricing.py

```diff
@@ -17,6 +23,99 @@
 log = logging.getLogger('pricing')
 
 
+G2_CPU_SKUS = [
+    # G2 Instance Core running in Iowa
+    'D18E-3563-415F',
+    # G2 Instance Core running in Columbus
+    'A312-C1EC-7F1A',
+    # Spot Preemptible G2 Instance Core running in Iowa
+    '59A6-9C49-BDDB',
+    # Spot Preemptible G2 Instance Core running in Columbus
+    'B56F-7169-FA2D',
+    # Spot Preemptible G2 Instance Core running in Las Vegas
+    'E7D9-CF86-B013',
+    # G2 Instance Core running in Virginia
+    'CF07-A027-C390',
+    # G2 Instance Core running in Los Angeles
+    'F81A-966F-28CD',
+    # Spot Preemptible G2 Instance Core running in Dallas
+    '36AA-92D4-D377',
+    # Spot Preemptible G2 Instance Core running in Los Angeles
+    'BA15-120B-D484',
+    # G2 Instance Core running in Salt Lake City
+    '321F-5134-70E7',
+    # Spot Preemptible G2 Instance Core running in Salt Lake City
+    '38F3-C0B3-4807',
+    # Spot Preemptible G2 Instance Core running in Virginia
+    '55E2-9EB8-778F',
+    # G2 Instance Core running in Dallas
+    'B911-29A9-DC06',
+    # G2 Instance Core running in Las Vegas
+    '29FF-9F02-F5CE',
+]
+
+G2_RAM_SKUS = [
+    # G2 Instance Ram running in Virginia
+    'A15F-B56F-97FE',
+    # G2 Instance Ram running in Iowa
```

Comment:
Can you explain to me why you had to hard code these?

Reply:
With this assert, you'll know you're picking up thing that aren't what you intend. However, it's not going to help with start missing updates to those intended products. It will just not see an update and not update the database. If you want that behavior, then you'll want to keep track of the SKUs you've seen and make sure that list is the set of all SKUs you expected. If the list is different, then don't throw an exception. Write a new log message with the missing updates `log.exception(f'missing g2 skus {expected_g2_skus.difference(seen_g2_skus)}')`. Throwing an exception will just kill the batch driver entirely, which is not what we want. We just want the alerts to start firing with that error message and then we can respond appropriately.

Reply:
OK, this might be a non-issue, I don't totally understand the pricing/billing code.

GCP products are identified by matching on the description (and other fields).  For example, from the code below:

```
            sku['description'].startswith("G2 Instance")
            or sku['description'].startswith("Spot Preemptible G2 Instance")
```

What happens if Google adds a new product that satisfies this condition?  Then we will generate duplicate resources (potentially with different prices).  There are already GCP products that have similar descriptions (e.g. Sole Tenant G2 instances).  Therefore, we decided to hardcode the expected GCP SKUs.  These are only used in assertions.

`_refresh_resources_from_retail_prices` would log an error in this case.  I'd be slightly uncomfortable to have the driver running with wrong/corrupted pricing and I would probably prefer a failure and be forced to figure out what's wrong.

Actually, there might not even be an error: if two GCP SKUs collide with the same resource, and one of them has a newer version (start date), the product might be inserted as a new version with no errors (but it should generate an error on the next restart).

If you'd prefer we can back this part out and have the new code match the existing style.  If this is a real concern, it might make sense to address it in a separate PR since it isn't GPU specific.

Reply:
I made a PR that will ensure that we don't update a product with the information of a different SKU then what was originally intended. Unfortunately, I am out on vacation the rest of the week. Feel free to adopt this and close https://github.com/hail-is/hail/pull/13607 if it will help you feel confident in your changes without needing the hard coded skus and the asserts. Otherwise, I can follow up with this change next week and you should keep the existing code that you are comfortable with including the hard coded skus and the asserts.